[5.1] Skip further date format checks if DateTime passed to asDateTime()

harrygr opened the pull request:
Despite the comments within the `asDateTime()` method saying to skip further checks if the `$value` is a `DateTime` object, the actual logic does not skip them.

Commit:
Skip further date format checks if DateTime passed to asDateTime() 

Despite the comments within the `asDateTime()` method saying to skip further checks if the `$value` is a `DateTime` object, the actual logic does not skip them.

franzliedke commented:
I think we've had this before. :+1:

The last `elseif` can also be turned into a simple `else`, no?

harrygr commented:
You could skip all the elseifs entirely actually and change them all to separate ifs as they each `return`. There are two opposite statements: `if($value instanceof DateTIme)` and the last one (after some other checks) `if(!$value instanceof DateTime)`. These two alone mean at least one of the checks will be satisfied.

My change failed a test though, but only on PHP 5.6. Can't figure out why.

harrygr commented:
I also feel that it might be better to change the restrictive `Carbon::createFromFormat($format, $value);` to a simple `Carbon::parse($value);`. This will allow other date formats to be passed and means the method doesn't need to know the date format.

franzliedke commented:
> My change failed a test though, but only on PHP 5.6. Can't figure out why.

Probably a timing issue. [These two Carbon instances](https://github.com/laravel/framework/blob/891ed3bc0329c61994985bec702e5818569decbf/tests/Database/DatabaseEloquentRelationTest.php#L33-34) should be the same ones, stored in a variable. They probably happened to have two different values for the second.

@GrahamCampbell Can you retrigger the Travis build to confirm?

Commit:
Tidy up unnecessary elseifs

All the conditinal checks `return` from the method so there is no need to have chained `elseifs`. Changed to separate `ifs` and removed the last check as it's implied if all the other checks are not satisfied.

GrahamCampbell commented on src/Illuminate/Database/Eloquent/Model.php:
```diff
@@ -2826,33 +2826,30 @@ protected function asDateTime($value)
 
-        return Carbon::instance($value);
+        return Carbon::createFromFormat($format, $value);
+        
```
no newline here please

Commit:
remove superfluous newline

lucasmichot commented on src/Illuminate/Database/Eloquent/Model.php:
```diff
@@ -2826,33 +2826,29 @@ protected function asDateTime($value)
         // these checks since they will be a waste of time, and hinder performance
         // when checking the field. We will just return the DateTime right away.
         if ($value instanceof DateTime) {
```
This should be added before
```php
if ($value instanceof Carbon) {
  return $value;
}
```

harrygr replied:
There is no need for that as if `$value` is an instance of Carbon it's also an instance of DateTime so will be caught by the first check.

lucasmichot replied:
Why would you instanciate a Carbon instance from another Carbon instance then ?

lucasmichot replied:
It would call uselessly `Carbon::instance`, `Carbon construct` and `Datetime construct` for an object that we know is already a `Carbon`

harrygr replied:
I agree it's not the most efficient approach, but it's what this method does before my PR anyway and results in one less if statement. I doubt it would have a meaningful impact on performance but yeah, it could be added.

lucasmichot replied:
Sure

Commit:
Added check for Carbon

As mentioned by @lucasmichot the `instanceof DateTime` check will also catch existing Carbon instances and reinstantiate them. This would be a waste of time given we already know it's a Carbon instance.

lucasmichot commented:
:+1:
